File: limix_inference/ep/liknorm/_liknorm.py
from __future__ import division

from numpy import ndarray

from . import _liknorm_ffi


def ptr(a):
    if isinstance(a, ndarray):
        return _liknorm_ffi.ffi.cast("double *", a.ctypes.data)
    return a


def create_liknorm(likelihood_name, nintervals):
    _liknorm_ffi.lib.initialize(nintervals)
    if likelihood_name.lower() == 'bernoulli':
        return BernoulliLikNorm()
    elif likelihood_name.lower() == 'binomial':
        return BinomialLikNorm()
    elif likelihood_name.lower() == 'poisson':
        return PoissonLikNorm()
    elif likelihood_name.lower() == 'exponential':
        return ExponentialLikNorm()
    raise ValueError


class BernoulliLikNorm(object):
    def __init__(self):
        self._moments = _liknorm_ffi.lib.binomial_moments

    def moments(self, phenotype, eta, tau, log_zeroth, mean, variance):
        outcome = phenotype.outcome
        from numpy import ones
        ntrials = ones((len(outcome), ))
        size = len(outcome)
        self._moments(size,
                      ptr(outcome),
                      ptr(ntrials),
                      ptr(eta),
                      ptr(tau), ptr(log_zeroth), ptr(mean), ptr(variance))

    def destroy(self):
        _liknorm_ffi.lib.destroy()


class BinomialLikNorm(object):
    def __init__(self):
        self._moments = _liknorm_ffi.lib.binomial_moments

    def moments(self, phenotype, eta, tau, log_zeroth, mean, variance):
        nsuccesses = phenotype.nsuccesses
        ntrials = phenotype.ntrials
        size = len(nsuccesses)
        _liknorm_ffi.lib.binomial_moments(size,
                                          ptr(nsuccesses),
                                          ptr(ntrials),
                                          ptr(eta),
                                          ptr(tau),
                                          ptr(log_zeroth),
                                          ptr(mean), ptr(variance))

    def destroy(self):
        _liknorm_ffi.lib.destroy()


class PoissonLikNorm(object):
    def moments(self, phenotype, eta, tau, log_zeroth, mean, variance):
        noccurrences = phenotype.noccurrences
        size = len(noccurrences)
        _liknorm_ffi.lib.poisson_moments(size,
                                         ptr(noccurrences),
                                         ptr(eta),
                                         ptr(tau),
                                         ptr(log_zeroth),
                                         ptr(mean), ptr(variance))

    def destroy(self):
        _liknorm_ffi.lib.destroy()


class ExponentialLikNorm(object):
    def moments(self, x, eta, tau, log_zeroth, mean, variance):
        size = len(x)
        _liknorm_ffi.lib.exponential_moments(size,
                                             ptr(x),
                                             ptr(eta),
                                             ptr(tau),
                                             ptr(log_zeroth),
                                             ptr(mean), ptr(variance))

    def destroy(self):
        _liknorm_ffi.lib.destroy()

File: limix_inference/ep/liknorm/clib/liknorm.c
#include "liknorm/liknorm.h"

static LikNormMachine *machine;

void initialize(int n) { machine = liknorm_create_machine(n); }

void destroy(void) { liknorm_destroy_machine(machine); }

void binomial_moments(double size, double *k, double *n, double *eta,
                      double *tau, double *log_zeroth, double *mean,
                      double *variance) {

  for (size_t i = 0; i < size; i++) {
    liknorm_set_normal(machine, tau[i], eta[i]);
    liknorm_set_binomial(machine, k[i], n[i]);
    liknorm_moments(machine, log_zeroth + i, mean + i, variance + i);
  }
}

void poisson_moments(double size, double *k, double *eta, double *tau,
                     double *log_zeroth, double *mean, double *variance) {

  for (size_t i = 0; i < size; i++) {
    liknorm_set_normal(machine, tau[i], eta[i]);
    liknorm_set_poisson(machine, k[i]);
    liknorm_moments(machine, log_zeroth + i, mean + i, variance + i);
  }
}

void exponential_moments(double size, double *x, double *eta, double *tau,
                         double *log_zeroth, double *mean, double *variance) {

  for (size_t i = 0; i < size; i++) {
    liknorm_set_normal(machine, tau[i], eta[i]);
    liknorm_set_exponential(machine, x[i]);
    liknorm_moments(machine, log_zeroth + i, mean + i, variance + i);
  }
}
